docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,124 @@
+
+## Alternatives considered
+
+### Aggregation in PromQL API clients
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+As this approach comes with the following disadvantages, it was not considered further:
+
+- Every PromQL API client needs to support the aggregation from various sources.
+
+- Queries that are written in PromQL can't be used without extra work across tenants.
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+### Multi tenant aggregation in the query frontends
+
+Another approach to multi tenant query federation could be achieved by aggregation of partial query results within the query frontend. For this a query needs to be split into sub queries per tenant and afterwards the partial results need reduced into the final result.
+
+The [astmapper] package goes down a similar approach, but it cannot parallelize all query types. Ideally multi-tenant query federation should support the full PromQL language and the algorithms necessary would differ per query funtion and operators used. This approach was deemed as a fairly complex way to achieve that tenant query federation.
+
+[astmapper]: https://github.com/cortexproject/cortex/blob/f0c81bb59bf202db820403812e8dabcb64347bfd/pkg/querier/astmapper/parallel.go#L27
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to be able to always identify the tenant correctly, queries using multiple tenants should inject a tenant label named `__tenant_id__` and its value containing the tenant ID into the results. A potentially existing label with the same name should be stored in a label with the prefix `original_`.
+
+Label selectors containing the tenant label should behave like any other labels. This can be achieved by selecting the tenants used in a multi tenant query.
+
+### Exposure of feature to the user
+
+#### Challenge
+
+The tenant ID is currently read from the `X-Scope-OrgID` HTTP header. The tenant ID has those [documented limitations][cortex_tenant_id] of values being allowed.
+
+[cortex_tenant_id]: https://cortexmetrics.io/docs/guides/limitations/#tenant-id-naming
+
+#### Proposal
+
+For the query path a user should be able to specify a `X-Scope-OrgID` header with multiple tenant IDs.  Multiple tenant IDs should then be propagating throughout out the system until it reaches the querier. The `Queryable` instance returned by the querier component, is wrapped by a `mergeQueryable`, which will aggregate the results from a `Queryable` per tenant and hence treated by the downstream components as a single tenant query.
```

Comment:
I am wondering if this is really the way to go, as it prohibits some scenarios.

I would suggest a different approach... define tenant grouping in a Cortex configuration. What I mean by that is that Cortex would have artificial tenant, which would actually map into group of tenants that are really queried.

This could be an overrideable per-tenant setting. By default, tenant would only map to itself, but it would be possible to say that tenant A is really querying data from “B,C,D” or even regex like “A-.*”.

This solves problem with having single user in the X-Scope-OrgID, and allows for more interesting scenarios. Eg. having “global” tenant that simply maps to “.*” (not expanding this recursively).

It also makes authentication easier — it’s not clear to me how authentication would work with multiple values in X-Scope-OrgID header, but it’s clear how it works for single value.

Is it too late to consider this alternative?

Reply:
> I would suggest a different approach... define tenant grouping in a Cortex configuration. What I mean by that is that Cortex would have artificial tenant, which would actually map into group of tenants that are really queried.

At some point I have considered using exactly that approach defining "virtual tenants". I was just a bit concerned by those factors:
- What happens if a virtual tenant uses paths were we don't have multi tenant support. (e.g. the read part) 
- Should we allow recursive virtual tenants?
- Changing of tenant mappings would require cache invalidations

> This could be an overrideable per-tenant setting. By default, tenant would only map to itself, but it would be possible to say that tenant A is really querying data from “B,C,D” or even regex like “A-.*”.

Are we having the ability to list all existing tenants easily or does that only work for tenants recently queried/ingested to?

> It also makes authentication easier — it’s not clear to me how authentication would work with multiple values in X-Scope-OrgID header, but it’s clear how it works for single value.

I agree that it makes the auth proxy simpler, but I think pretty much the same logic will be part of cortex.

> Is it too late to consider this alternative?

No, I think it's better we have consensus here. 

@pstibrany not too sure how busy you are today, but maybe we can schedule a short 1-1 meeting before the next weekly and talk through it?

Reply:
> What happens if a virtual tenant uses paths were we don't have multi tenant support. (e.g. the read part)

On read path, we would have support. On write path, we don't expect any data for such user. We could add a check to disallow pushing data with mapping enabled.

> Should we allow recursive virtual tenants?

I think that would make it too complicated.

> Changing of tenant mappings would require cache invalidations

Very good point! We can solve it by storing the actual used mapping (=list of tenants) into cache and so that we can verify that cached data is still correct.

> Are we having the ability to list all existing tenants easily or does that only work for tenants recently queried/ingested to?

When using blocks, we do. When using chunks, we don't (= no regex based mapping, but explicit listing is still possible).

> I agree that it makes the auth proxy simpler, but I think pretty much the same logic will be part of cortex.

Cortex doesn't do anything other than checking that X-Scope-OrgId is not empty, and existing proxies / gateways can treat virtual tenants as normal users.

Reply:
After internal discussion, the main difference between this proposal and my suggestion is about where the mapping is stored (externally in case of this proposal, or within Cortex in my suggestion).

I think implementing this proposal will be a good step forward, and if we ever want to have mapping stored within Cortex, it should be easy to add in the future.